README.md
# DateByExample

Provides a Ruby implementation of Go-style date formatting by example

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'date_by_example'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install date_by_example

## Usage

### Basic usage

This gem adds a `by_example` method to `Date`, `Time`, and `DateTime`.

The argument to this method is the formatted version of a reference date.

For example:

```
> d = Date.today
> d.by_example("2006-01-02")   => "2018-06-11"

> t = DateTime.now
> t.by_example("Jan 02 2006 @ 3:04 pm")  => "Jun 11 2018 @ 3:52 pm"

> t = Time.now
> t.by_example("Jan 02 2006 @ 3:04 pm")  => "Jun 11 2018 @ 3:53 pm"
```

The reference date is Jan 2, 2006 at 3:04:05 PM MST, which can be
remembered as 1 2 3 4 5 6 -7.

I realize that remembering the reference date is not the easy part,
but I think it's easier than remembering all the strftime format
strings

### Supported formats

* Year: 2000 and 00 (but not Ruby's %c for the century)
* Month: January, JANUARY, Jan, JAN, 1, and 01, but not " 1" for space padding
* Day: 2, 02, 002 (for day of year), Monday, MONDAY, Mon, MON, but not " 2" for space padding
* Hour: 15, 3, pm, PM
* Minute: 04
* Second: 05, 05.000 for milliseconds and 05.000000 and 05.000000000
* Time Zone: -7000, -07:00, -07:00:00, MST

### Unsupported Formats

This does not support Ruby's week based formats

### Alternate usage

The slow part of using this gem is generating the format string from
the example, if you want to reuse a format multiple times, you can do so
by creating and reusing an `ExampleFormatter`

```
> ef = ExampleFormatter.new("01/02/06")
> ef.format(Date.today)   => "06/11/18"
> ef.format(Time.now)     => "06/11/18"
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/date_by_example. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the DateByExample project’s codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/[USERNAME]/date_by_example/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
## Version 0.1.1 6/22/2018

* Adds "%l" for single digit hours


spec/spec_helper.rb
require "bundler/setup"
require "date_by_example"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/date_by_example_spec.rb
RSpec.describe DateByExample do
  it "has a version number" do
    expect(DateByExample::VERSION).not_to be nil
  end
end


spec/example_formatter_spec.rb
RSpec.describe ExampleFormatter do

  def format_string_for(str)
    ExampleFormatter.new(str).format_string
  end

  def format_date(date, format)
    ExampleFormatter.new(format).format(date)
  end

  describe "years" do

    specify { expect(format_string_for("2006")).to eq("%Y") }
    specify { expect(format_string_for("06")).to eq("%y") }
    # ruby allows "%C for century, but I think we can't allow that here
    # because 20 is be ambiguous with 201

  end

  describe "months" do

    specify { expect(format_string_for("01")).to eq("%m") }
    # ruby has "%_m" for space padding, but we're not doing that
    specify { expect(format_string_for("1")).to eq("%-m") }
    specify { expect(format_string_for("January")).to eq("%B") }
    specify { expect(format_string_for("JANUARY")).to eq("%^B") }
    specify { expect(format_string_for("Jan")).to eq("%b") }
    specify { expect(format_string_for("JAN")).to eq("%^b") }

  end

  describe "day of month" do

    specify { expect(format_string_for("02")).to eq("%d") }
    # ruby has "%e" for space padding, but we're not doing that
    specify { expect(format_string_for("2")).to eq("%-e") }
    specify { expect(format_string_for("002")).to eq("%j") }
    specify { expect(format_string_for("Monday")).to eq("%A") }
    specify { expect(format_string_for("MONDAY")).to eq("%^A") }
    specify { expect(format_string_for("Mon")).to eq("%a") }
    specify { expect(format_string_for("MON")).to eq("%^a") }
    # ruby's numerical days of the week %u, %w probably can't be done

  end

  describe "hours" do
    specify { expect(format_string_for("15")).to eq("%H") }
    # ruby has %k for blank padded 24 hour clock but I can't do that here
    specify { expect(format_string_for("03")).to eq("%I") }
    specify { expect(format_string_for("3")).to eq("%l") }
    # ruby has "%l" for space padding, but we're not doing that
    specify { expect(format_string_for("PM")).to eq("%p") }
    specify { expect(format_string_for("pm")).to eq("%P") }
  end

  describe "minutes and seconds" do
    specify { expect(format_string_for("04")).to eq("%M") }
    specify { expect(format_string_for("05")).to eq("%S") }
    specify { expect(format_string_for(".000")).to eq(".%L") }
    specify { expect(format_string_for(".000000")).to eq(".%6N") }
    specify { expect(format_string_for(".000000000")).to eq(".9N") }
  end

  describe "timezone" do
    specify { expect(format_string_for("-7000")).to eq("%z") }
    specify { expect(format_string_for("-07:00")).to eq("%:z") }
    specify { expect(format_string_for("-07:00:00")).to eq("%::z") }
    specify { expect(format_string_for("MST")).to eq("%Z") }
  end

  # other things that can't be done here
  # %G -- week-based year
  # %g -- last two digits of week based year
  # %V -- week number of week based year
  # %U -- week number of year (sunday based)
  # %W -- week number of the year (monday based)
  # %s -- seconds since the epoch

  describe "can do date formats" do
    let(:date) { DateTime.new(2018, 6, 10, 13, 15) }

    specify { expect(format_string_for("Jan 2, 2006")).to eq("%b %-e, %Y") }
    specify { expect(format_date(date, "Jan 2, 2006")).to eq("Jun 10, 2018") }
    specify { expect(format_date(date, "01/02/06")).to eq("06/10/18") }
    specify { expect(format_string_for("2 Jan 06 15:04")).to eq("%-e %b %y %H:%M") }
    specify { expect(format_date(date, "2 Jan 06 15:04")).to eq("10 Jun 18 13:15") }
    specify { expect(date.by_example("Jan 2, 2006")).to eq("Jun 10, 2018") }
    specify { expect(date.by_example("January 02, 2006 @ 03:04pm")).to eq("June 10, 2018 @ 01:15pm") }
  end

end


lib/date_by_example/example_formatter.rb
class ExampleFormatter
  attr_accessor :reference

  def initialize(reference)
    @reference = reference
    @format_string = nil
  end

  FORMATS = {
    ".000000000" => ".9N",
    "-07:00:00" => "%::z",
    ".000000" => ".%6N",
    "January" => "%B",
    "JANUARY" => "%^B",
    "Monday" => "%A",
    "MONDAY" => "%^A",
    "-07:00" => "%:z",
    "-7000" => "%z",
    "2006" => "%Y",
    ".000" => ".%L",
    "002" => "%j",
    "Jan" => "%b",
    "JAN" => "%^b",
    "Mon" => "%a",
    "MON" => "%^a",
    "MST" => "%Z",
    "15" => "%H",
    "06" => "%y",
    "01" => "%m",
    "02" => "%d",
    "03" => "%I",
    "PM" => "%p",
    "pm" => "%P",
    "04" => "%M",
    "05" => "%S",
    "1" => "%-m",
    "2" => "%-e",
    "3" => "%l"}.freeze

  FORMAT_MATCHER = Regexp.union(FORMATS.keys)

  def format_string
    @format_string ||= reference.gsub(FORMAT_MATCHER, FORMATS)
  end

  def format(date)
    date.strftime(format_string)
  end
end
